### PR TITLE
InspectorControls: Wrap block support slots in ToolsPanel

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
Depends on:
- https://github.com/WordPress/gutenberg/pull/34632

Related: 
- https://github.com/WordPress/gutenberg/pull/34069
- ~https://github.com/WordPress/gutenberg/pull/34063~
- https://github.com/WordPress/gutenberg/pull/34065

## Description

- Extends the work done in https://github.com/WordPress/gutenberg/pull/34069 to provide groups in `InspectorControls`.
- Adds a slot for the dimensions block support 
- Adds the new slot into the block inspector
- Wraps block support InspectorControl groups in a ToolsPanel. 
    - Does this via new component that handles the generation of `resetAll` function, getting block attributes from the store etc.
- Updates dimensions block support hook to inject its padding and margin controls via the hook

## How has this been tested?

1. Enable padding and margin support in theme.json
2. Opt into margin support for the group block ( we need two to properly test )
3. Edit a post, add a group block and select it
4. Confirm that the dimensions panel appears in the sidebar
5. Toggle the padding and margin controls on and off and ensure they display correctly
6. Set values for both padding and margin. Then test resetting them individually via the panel's menu
7. Set both values again before this time selecting "reset all" from the panel's menu
8. Ensure both values have been reset and the block has updated correctly. 

## Types of changes
New feature. Enhancement.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/block-editor/src/components/block-inspector/index.js
Comment (youknowriad) [inline packages/block-editor/src/components/block-inspector/index.js]: @mtias mentioned that there are still some design explorations pending that might impact the API a bit (adding tabs and groups). So we might want to start with `__experimentalGroup` to be able to change it later if needed.

**Review by youknowriad — COMMENTED** [review]
I like this approach personally. But you worked on both, you'd have a better insight here, what are your current thoughts?

**Comment by gziolo** [comment]
Just letting you know that applied some breaking changes to https://github.com/WordPress/gutenberg/pull/34069 and therefore to the base branch used here. The only important change is that `group` was marked as experimental based on the feedback received, so it's now `__experimentalGroup`.

File: packages/block-editor/src/components/block-inspector/index.js
Comment (aaronrobertshaw) [reply packages/block-editor/src/components/block-inspector/index.js]: ✅ Done. Thanks.

File: packages/block-editor/src/components/inspector-controls/slot.js
Comment (andrewserong) [inline packages/block-editor/src/components/inspector-controls/slot.js]: Is it worth us using a shared constant between these strings and the ones in `block-support-tools-panel` to ensure they match?

**Comment by aaronrobertshaw** [comment]
> I like this approach personally. But you worked on both, you'd have a better insight here, what are your current thoughts?

This approach will make the block support hooks a little cleaner, reducing some boilerplate code. For the immediate uses planned, I like this approach as well. 

If a scenario does arise in the future where a block support panel needs to do something more than just reset attributes, we can cross that bridge then.

While creating this PR I did run into some odd behaviour resetting the block attributes via `updateBlockAttributes` instead of the `setAttributes` passed through props. That ultimately led to tightening up the `ToolsPanel` which is another win.

I'll close the original PR in favour of this one.

**Review by andrewserong — COMMENTED** [review]
This is testing well for me @aaronrobertshaw, I like this approach, and nice work fixing up the `resetAll` behaviour!

I was wondering if it's worth us adding a `constants.js` file to limit using string literals in the `__experimentalGroup` prop for `dimensions`, `border`, `color`, `typography` etc? It's definitely something we could look into in a follow-up instead (so not blocking for this PR), but just to help avoid accidental typos in subsequent PRs, which are sometimes tricky to debug. This comment might be more applicable to #34069 of course, so I'll add a comment over there, too.

Otherwise, looks great to me!

File: packages/block-editor/src/components/inspector-controls/block-support-tools-panel.js
Comment (andrewserong) [inline packages/block-editor/src/components/inspector-controls/block-support-tools-panel.js]: I was wondering if we needed optional chaining here. We probably don't because by the time we're inside this function, there's already been a check that the `group` exists. But, to guard against accidentally missing a group in the list of labels, we could potentially use constants for the `labels` keys. Not a blocker for this PR since it's all working well, just thinking about making it easy make changes in follow-ups 🙂

Comment (aaronrobertshaw) [reply packages/block-editor/src/components/inspector-controls/block-support-tools-panel.js]: > We probably don't because by the time we're inside this function, there's already been a check that the group exists

That was my thinking. I'm happy to tighten this up though.

File: packages/block-editor/src/components/inspector-controls/slot.js
Comment (andrewserong) [reply packages/block-editor/src/components/inspector-controls/slot.js]: That's looking good now, thanks for tightening this up! 👍

Comment (aaronrobertshaw) [reply packages/block-editor/src/components/inspector-controls/slot.js]: It definitely could be. 

For now, I've opted to only tweak this such that the labels object is used for this check. I think the benefit of the shared constants would come in if they are adopted when [defining the groups](https://github.com/WordPress/gutenberg/blob/7c23fd8d87361eee366797998cc36a9c2f40e17a/packages/block-editor/src/components/inspector-controls/groups.js) and consuming the API:
e.g. `<InspectorControls group={ GROUP_DIMENSIONS }>`

I'm happy to add the constants in a follow up PR to this and https://github.com/WordPress/gutenberg/pull/34069

File: packages/block-editor/src/components/inspector-controls/block-support-tools-panel.js
Comment (aaronrobertshaw) [reply packages/block-editor/src/components/inspector-controls/block-support-tools-panel.js]: I've added the optional chaining as suggested and replied in relation to the constants in this https://github.com/WordPress/gutenberg/pull/34157#discussion_r692687902.

File: packages/block-editor/src/components/inspector-controls/slot.js
Comment (andrewserong) [reply packages/block-editor/src/components/inspector-controls/slot.js]: Sounds good to me, the way you've combined them addresses the initial concern I had, and agreed, the individual / shared constants will be more useful when defining the groups and consuming the API as you suggest 👍. I'm happy to leave any other changes to follow-ups!

File: packages/block-editor/src/components/inspector-controls/groups.js
Comment (gziolo) [inline packages/block-editor/src/components/inspector-controls/groups.js]: We probably could use the following form to align with `BlockControls`:
```suggestion
	'InspectorControlsDimensions'
```

Screen Shot 2021-08-24 at 09 55 07

The existing exception are mostly for backward-compatibility like `BlockFormatControls` or `InspectorAdvancedControls`. Well, for everything else that was never used in the code as a string we can go with a consistent pattern we agree upon.

Comment (aaronrobertshaw) [reply packages/block-editor/src/components/inspector-controls/groups.js]: Sounds good to me.

✅  Done.

**Review by andrewserong — APPROVED** [review]
Good work @aaronrobertshaw, this is testing well for me, and nice idea using the `panelId` to ensure that injected `ToolsPanelItem`s are connected to the correct `ToolsPanel`. It works well in testing, and as we discussed in https://github.com/WordPress/gutenberg/pull/34065#issuecomment-905978146, once the `__experimentalGroup` prop stabilises, it'd be good for us to update the InspectorControls documentation with examples on how to use the ToolsPanel item and hook it up to the right ToolsPanel via the clientId. Since this is still experimental, I think it's fine to delay getting the docs in as we're likely to change implementation details over time.

The behaviour in this PR and https://github.com/WordPress/gutenberg/pull/34065 feels solid to me in testing, and I think it includes some good pragmatic workarounds for the edge cases we encountered, so I'm giving it the ✅, but let's see what others think about the `panelId` idea, too.

**Comment by aaronrobertshaw** [comment]
@gziolo or @youknowriad do you have any other thoughts or concerns I need to address on this PR? It would be great to sneak it, and the related cover block PR, into this release if possible.

In particular, I thought you might have some better ideas around how I solved the issues where switching between block selections still had previous fills hanging around which caused incorrect panel items to be registered.

File: packages/block-editor/src/components/inspector-controls/slot.js
Comment (gziolo) [inline packages/block-editor/src/components/inspector-controls/slot.js]: This check is confusing. It will work only for „dimensions” group. For other values, it will never reach this place, because there is no slot defined. Wouldn’t it be more readable if we would move that to the place where it is actually used?

Comment (aaronrobertshaw) [reply packages/block-editor/src/components/inspector-controls/slot.js]: The intention here was that after this is done, the other block supports such as border, color and typography would have slots defined as they are moved to the new ToolsPanel.

Moving generation of the ToolsPanel here instead of in the block supports was a suggestion made on the original PR: https://github.com/WordPress/gutenberg/pull/34063. The idea being then that the block supports all used their respective slots to inject their controls, allowing us to not duplicate the code generating the ToolsPanel for each block support.

Do you think it is better to move this to where the [slot is used in the Block Inspector](https://github.com/WordPress/gutenberg/pull/34157/files#diff-068538657061c47f0a58cfc331259bbeabe11bcc0c576a0cc496d262e5b0038cR132)?  

We'd end up with some extra duplication of checks for there being no fills to prevent rendering an empty panel as well as then needing to wrap each block support slot individually. That nearly brings us back to the original PR.

Comment (gziolo) [reply packages/block-editor/src/components/inspector-controls/slot.js]: There are many other ways you could code it avoiding `BLOCK_SUPPORT_GROUPS`.

```js
<InspectorControls.Slot
 	__experimentalGroup="dimensions"
	label={ __( 'Dimensions options' ) }
	title={ __( 'Dimensions' ) }
 	bubblesVirtually={ false }
 />
```

You could then wrap the slot with `<BlockSupportToolsPanel title={ title } label={ label } />` when the `title` prop is passed if you prefer to avoid the dance with checking whether fills need to be rendered.

Feel free to wait for feedback from other reviewers and see what they think. I'm mostly concerned about two things:
- the indirection that `BLOCK_SUPPORT_GROUPS` creates where a group prop magically transforms into other props 2 levels deeper
- the missing slots for other block support groups – it seems to be temporary though

Comment (aaronrobertshaw) [reply packages/block-editor/src/components/inspector-controls/slot.js]: Ah ok. Thanks 👍 

Avoiding that indirection sounds good. I've updated this PR along the lines of your example. The prop has been kept as `header` instead of `title` so as to match the `ToolsPanel` given it was switched to that to avoid some accessibility concerns/confusion over the prop in that component.

Regarding the "missing slots", I think we can still hold off on those and add them as needed. Once this lands though I'll definitely update the current PRs switching panels for those block supports, to add and use an appropriate group slot (https://github.com/WordPress/gutenberg/pull/33743, https://github.com/WordPress/gutenberg/pull/34027, https://github.com/WordPress/gutenberg/pull/33744).

File: packages/block-editor/src/components/inspector-controls/block-support-tools-panel.js
Comment (gziolo) [inline packages/block-editor/src/components/inspector-controls/block-support-tools-panel.js:18]: We are exploring ways to absorb parent controls as proposed in https://github.com/WordPress/gutenberg/issues/26313. The fact that this panel depends on the currently selected block would prevent it from being usable in such context. Maybe it's fine because we might never get to the point where you could expose those controls from the parent to child blocks. It might be also not hard to refactor it when needed and pass the parent client id through the slot when necessary. I mostly wanted to raise awareness. I don't think we need to worry that much about it in the initial implementation, but we should keep that in mind that this panel doesn't necessarily need to be concerned only about the currently selected block.

File: packages/block-editor/src/components/block-inspector/index.js
Comment (youknowriad) [inline packages/block-editor/src/components/block-inspector/index.js]: What's the difference between "label" and "header". Shouldn't be just the same thing?

Comment (aaronrobertshaw) [reply packages/block-editor/src/components/block-inspector/index.js]: The `header` is the panel header for the ToolsPanel. The `label` is used for the dropdown menu's label.

As the use of the `ToolsPanel` component becomes more varied, I can see cases where the header may wish to contain HTML markup, e.g. color swatch indicators, or the label for the panel's menu needs to be more than just `<header> options`.

**Review by glendaviesnz — APPROVED** [review]
Worked as advertised for me when tested with group block:

✅ - Padding displayed by default
✅  - Margin could be toggled on
✅  - Both could be updated as one setting, or 4 edges set independently, and changes reflected in editor and front end
✅  - Removing from the panel reset values, as did using the `Reset all` option

File: packages/block-editor/src/components/block-inspector/index.js
Comment (mtias) [reply packages/block-editor/src/components/block-inspector/index.js]: Do we have examples of the header of the Tools Panel being more varied design-wise? This might not prove to be a necessary affordance and might be good to tighten it first. (The label on the dropdown could also be better as just "Options" or nothing at all.)

Comment (aaronrobertshaw) [reply packages/block-editor/src/components/block-inspector/index.js]: > Do we have examples of the header of the Tools Panel being more varied design-wise? 

I mistakenly took it that the color block support panel when switched to the tools panel would continue to display its color swatches in the header as it does now.  Revisiting the designs in https://github.com/WordPress/gutenberg/issues/27331 shows the color panel without swatches. So these props could be merged.

> The label on the dropdown could also be better as just "Options" or nothing at all.

The `DropdownMenu` component readme states the `label` is required for accessibility purposes so if not merging `header` and `label` perhaps your "Options" suggestion better than nothing at all?

Would it be acceptable to make these changes to the `ToolsPanel` component in a separate PR? 

Either way, I can update this tomorrow.

Comment (mtias) [reply packages/block-editor/src/components/block-inspector/index.js]: Yes, let's merge the props for now and we can see if there's a need to extend it later. Without a collapsed state, there's not much need for previews like the color swatch at the moment.

Comment (aaronrobertshaw) [reply packages/block-editor/src/components/block-inspector/index.js]: The `header` and `label` props for the `ToolsPanel` have been merged ( https://github.com/WordPress/gutenberg/pull/34157/commits/ad5fb02c6ffec757b5fe2beb9ac026cff06746a9) . I went with `label` as that seems to better reflect how the text is used and the limitations imposed upon it.

✅ ToolsPanel tests are passing
✅ Post editor panel for dimensions support is working
✅ Global styles panel in site editor is okay
✅ Storybook entry for tools panel was updated as well

Thanks for the help refining this. Hopefully, this is getting close to being ready to land now.

**Review by andrewserong — APPROVED** [review]
I just gave this another test after the rebase, and it's still working nicely:

✅ Tested padding in Group and Column blocks
✅ Tested opting-in to margin in the Group block
✅ Tested opting-in to blockGap in the Buttons block

All worked as expected in the post and site editors (and on the front end), resetting individual values and all values, and clicking between similar blocks.

LGTM!

**Comment by aaronrobertshaw** [comment]
Thanks for taking another look and reviewing this @andrewserong 👍 

Now that previous suggestions have been implemented, including merging the header/label props on the `ToolsPanel`, I think this is close to landing. Any issues around absorption of parent controls are planned to be iterated on separately.

Anything else you think needs to be addressed @gziolo or @youknowriad?

File: packages/block-editor/src/components/block-inspector/index.js
Comment (youknowriad) [inline packages/block-editor/src/components/block-inspector/index.js:134]: Why do we use `bubblesVirtuall = { false }`. Ideally we always use true as we were thinking of deprecating the other approach ultimately.

Comment (aaronrobertshaw) [reply packages/block-editor/src/components/block-inspector/index.js:134]: I found when `bubblesVirtually` was `true` it interfered with the registration of panel items via the `ToolsPanel`'s context.

Comment (gziolo) [reply packages/block-editor/src/components/block-inspector/index.js:134]: It's a know limitation that you need to propagate the context when using slots that bubble virtually. An example when the context is passed down through `fillProps` is for the block controls:

https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/block-controls/slot.js

@diegohaz might remember better the implementation details. I'm not sure how much it differs from the use case you have, but we should try to generalize the solution if possible.

Comment (aaronrobertshaw) [reply packages/block-editor/src/components/block-inspector/index.js:134]: Thanks @gziolo, the example you gave has helped get me most of the way to fixing this up.

The only real difference I see in this use case is that the `ToolsPanelContext` isn't being created until the `ToolsPanel` is [when it wraps the Slot](https://github.com/WordPress/gutenberg/pull/34157/files#diff-d54bdb7093990ff38b7f03adf368284bb43cb765ee9100b938c7b5db54e6ba8bR33-R35). I believe this means we need to wrap the Slot again so that component can grab the context and pass it through the `fillProps`.

I tried a HoC approach here but couldn't get that working. Instead, I have a [regular component handling this](https://github.com/WordPress/gutenberg/commit/1659b0ef54e35dadcea995dd6df053a21f96af3a). It almost "works" except that it introduces a rogue `<div>` wrapping all the fills. This breaks the layout of the `ToolsPanel`.  I haven't been able to put my finger on where I've gone wrong here. 

Any help you can offer here would be greatly appreciated 🙏  

Also, would this be something I could continue to iterate on in a separate PR?

Comment (gziolo) [reply packages/block-editor/src/components/block-inspector/index.js:134]: If everything else is working then you can tackle it separately. It's Important to find a good solution here because the same approach is going to be replicated in other places 😄

Comment (diegohaz) [reply packages/block-editor/src/components/block-inspector/index.js:134]: > The only real difference I see in this use case is that the `ToolsPanelContext` isn't being created until the `ToolsPanel` is [when it wraps the Slot](https://github.com/WordPress/gutenberg/pull/34157/files#diff-d54bdb7093990ff38b7f03adf368284bb43cb765ee9100b938c7b5db54e6ba8bR33-R35). I believe this means we need to wrap the Slot again so that component can grab the context and pass it through the `fillProps`.

How is the component tree structured? Won't the Slot component be rendered inside ToolsPanel, in which case it would have access to the parent context?

Comment (aaronrobertshaw) [reply packages/block-editor/src/components/block-inspector/index.js:134]: > If everything else is working then you can tackle it separately. It's Important to find a good solution here because the same approach is going to be replicated in other places 😄

This PR's current state is working well according to the most recent reviews and testing. Myself, @glendaviesnz and @andrewserong have all put it through its paces.

If you are comfortable with this being merged, including the `bubblesVirtually={ false }`, I'll definitely continue iterating on this separately until that good solution is found.

Comment (aaronrobertshaw) [reply packages/block-editor/src/components/block-inspector/index.js:134]: This is a bit out of my normal wheelhouse so please excuse my lack of understanding around this.

> How is the component tree structured? Won't the Slot component be rendered inside ToolsPanel, in which case it would have access to the parent context?

That was what I was originally expecting but it only worked when preventing virtual bubbling.

The ToolsPanel creates the context and wraps its children in the context provider. The slot is rendered within that. The panel items are added via that slot. Those panel items need to access the ToolsPanelContext to register themselves with the panel. 

My current understanding is that the slot will have access to the context but the context has to be passed through `fillProps` so that we can re-create the context provider to wrap the fills thereby allowing the panel items access to the context again.

I hope that makes sense!

File: packages/components/src/tools-panel/tools-panel-item/hook.js
Comment (ciampo) [inline packages/components/src/tools-panel/tools-panel-item/hook.js]: Should `isResetting` be added to the dependencies of the `useEffect` hook?

File: packages/components/src/tools-panel/tools-panel/hook.js
Comment (ciampo) [inline packages/components/src/tools-panel/tools-panel/hook.js]: Without any dependencies listed, this `useEffect` will be executed for every render. Is that intended? If yes, to what purpose (and why would `useEffect` be necessary)?

**Review by ciampo — COMMENTED** [review]
Sorry for the late review, but I've only just realised this PR was in the works!

In general, it'd be good to keep the components squad in the loop every time a `@wordpress/components` component gets edited — and especially when the APIs are modified.

Also, it'd be great if in the future we could avoid mixing block editor and `@wordpress/components` changes within the same PR

**Comment by aaronrobertshaw** [comment]
> Sorry for the late review, but I've only just realised this PR was in the works!

It's very much appreciated. My apologies for missing the components label on this PR and specific components squad members from the reviewers. 

> Also, it'd be great if in the future we could avoid mixing block editor and @wordpress/components changes within the same PR

Understood. The changes here evolved from issues uncovered when attempting to implement the use of the ToolsPanel around the new grouped InspectorControls slots. These were also in a separate branch when this PR was first created, so splintering this again at that stage was problematic. 

With the majority of those issues debugged and addressed. I'd be happy to split out the ToolsPanel updates from this PR if it will help move this all forward. I'll also be more aware of isolating component changes moving forward. Thanks.

File: packages/components/src/tools-panel/tools-panel-item/hook.js
Comment (aaronrobertshaw) [reply packages/components/src/tools-panel/tools-panel-item/hook.js]: ✅ Fixed. Thanks.

File: packages/components/src/tools-panel/tools-panel/hook.js
Comment (aaronrobertshaw) [reply packages/components/src/tools-panel/tools-panel/hook.js]: ✅ Fixed. Thanks.

**Comment by aaronrobertshaw** [comment]
The updates to the `ToolsPanel` required to facilitate support for injecting panel items via SlotFills have been split into a separate PR https://github.com/WordPress/gutenberg/pull/34632. 

This PR now targets https://github.com/WordPress/gutenberg/pull/34632 as its base.

**Comment by gziolo** [comment]
> In general, it'd be good to keep the components squad in the loop every time a @wordpress/components component gets edited — and especially when the APIs are modified.

I'm afraid it isn't a realistic expectation in the project that has over 800 contributors. There isn't even a formal WordPress team for `@wordpress/components` that the community could learn about. Fortuntately there are ways to get better overview about all the changes proposed in PRs:
- everyone willing to review changed to `packages/components` folder can add their GitHub handles to the [CODOWNERS file](https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/creating-a-repository-on-github/about-code-owners) to get notification when PR gets opened that modified files in the folder:
  https://github.com/WordPress/gutenberg/blob/e3591aa86bfc4fe02da616ef50e722504ba8b3d8/.github/CODEOWNERS#L79
- we discussed it in the past, but we could build GitHub action that automatically applies labels based on the type of modified files
- more time-consuming approach is to go through the list of commits and opened PRs and scan for related changed on a regular schedule (week or month)
- even when something get merged, it doesn't mean it can't be reverted or updated

> Also, it'd be great if in the future we could avoid mixing block editor and @wordpress/components changes within the same PR

Can you expand on this point? I don't think I agree on principle with this limitation, but many I don't know the full story.

**Comment by gziolo** [comment]
It looks like we need the following steps to move forward with this PR:
1. https://github.com/WordPress/gutenberg/pull/34632 waits for review from @ciampo, once it's approved we can merge and rebase this PR with `trunk`
2. We seem to agree that this PR can land as is after point 1 happens.
3. I discussed with @aaronrobertshaw that he will open a follow-up PR after point 2 happens. The scope will be limited to the changes to the SlotFill implementation as discussed in https://github.com/WordPress/gutenberg/pull/34157#discussion_r702680043. I'm happy to help with finding a working solution. @diegohaz and everyone else who find it intriguing you are all welcome to join efforts 😄

**Comment by ciampo** [comment]
> Can you expand on this point? I don't think I agree on principle with this limitation, but many I don't know the full story.

There are a few reasons behind my previous suggestion about keeping changes to the `@wordpress/components` package separate from the rest of the changes.

Firstly, splitting PRs into smaller chunks makes them easier to review and, in particular, isolating changes to the `@wordpress/components` package makes it easier to search/track/understand them at a later time (for example, for tracking down the root cause of a bug and issuing a fix).

But most importantly, the `@wordpress/components` package aims at providing a base for building UIs in Gutenberg (and beyond). Changes made to this package need to be considered more carefully compared to standard "product" work, since they will have repercussions for all of the users of this package (and not only the Gutenberg repo). This is even more important when considering how much we value not introducing breaking changes.

The `@wordpress/components` package still contains legacy components that are not always consistent with each other and that feel like "copy-pasted app components", lacking that level of polish and thoughtfulness that should be part of a good components library. Our recent efforts, like [updating the contributing guidelines](https://github.com/WordPress/gutenberg/pull/33960) and [investigating APIs consistency](https://github.com/WordPress/gutenberg/issues/33391), aimed at improving the situation in these areas.

With this perspective in mind, I don't see this being a limitation — rather an improvement in how we collaborate on the `@wordpress/components` package, which should help to increase its overall quality and consistency.

This should hopefully also give more context on the other point about involving the components squad:

> I'm afraid it isn't a realistic expectation in the project that has over 800 contributors. There isn't even a formal WordPress team for @wordpress/components that the community could learn about.

I agree that this is not realistic, and I quite like some of the ideas that you suggested. Ultimately we need to think about what's the best and most scalable solution.

**Comment by gziolo** [comment]
> Firstly, splitting PRs into smaller chunks makes them easier to review and, in particular, isolating changes to the @wordpress/components package makes it easier to search/track/understand them at a later time (for example, for tracking down the root cause of a bug and issuing a fix).

It depends on the perspective. From the development perspective it's easier to operate on a single branch as you can better drive the API decision which you can see in the discussions that happened in this PR. Splitting the work into smaller PRs to land changes seems like a good middle ground here if that helps to review the changes to `@wordpress/components` as discussed above.

> The @wordpress/components package still contains legacy components that are not always consistent with each other and that feel like "copy-pasted app components", lacking that level of polish and thoughtfulness that should be part of a good components library.

100% agree. This is exactly how this collection of components was born and evolved. That's why G2 components project has been started to address all the pain points you outlined and I personally love all the efforts towards making APIs unified and better organized 😄

**Comment by andrewserong** [comment]
Just gave this is another smoke test, and it's working nicely for me with the `useEffect` added back in for clearing the `isResetting` value. Looks like that now ensures that clearing the value happens after the render is completed, instead of part way through 👍

**Pull request merged**
